docs/rfcs/0003-rfc-controller_refactor.md

```diff
@@ -0,0 +1,115 @@
+
+One exception to the rule is `ConnectionHealth::ConnectionStability::Unstable`. The connection health object periodically sends pings and measures the received responses to assess whether or not the network connectivity is stable. To continue notifying users about their unstable network, we will continue to use this state within `ConnectionManager`. 
+
+Beyond the connectivity checks, the Controller also manages how much time has elapsed since the VPN was toggled on. Having this logic in the controller bloats the code without any gains, so that code will be moved out of `Controller` and instead a simple `Timer` object will encapsulates the related logic. The timer starts when the VPN enters State On since doing the initial checks may take a few seconds that we don't need to account for.
+
+Because this is all part of a unified connectivity experience, we do not want to release bits and pieces of this until everything is implemented, tested, and ready to go, this means that all implementation will live behind a feature flag until we are ready to test and release.
+
+
+## Implementation Plan
+Introduce a new `ConnectionManager` component that is activated when the client is launched and connects to `networkWatcher` in case anything changes. This new component will manage the following work:
+- Internet probe: assess if there is a network connection.
+- Firewall: assess if firewall is limiting network connectivity.
+- Captive portal check: determine uf the network is behind a captive portal.
+- Server probe: upon activation and then priodically check if there are available servers within user's selection location.
+- Subscription status: ensure that user has an active subscription, and if not, show the "No Subscription Found" screen.
+
+The logic to calculate elapsed time since the VPN was activated will also be moved out of the Controller and into a new `Timer` component. This bit of work will likely require a separate RFC. 
+
+
+The `ConnectionManager` will have two state machines:
+1. `VPNClientState` to determine if the VPN client is off, connecting, or on. 
+2. `ConnectionState` to keep track of which connection related task the VPN is performing at any given point. 
```

Comment:
Does this only correspond to active connection/disconnection states? If `VPNClientState` is on or off, what is the `ConnectionState`?

Reply:
Good catch, updated the ConnectionState values

Reply:
Thanks! And so `VPNClientState` is derived from `ConnectionState` right? It shouldn't be a variable that can be set, right? Don't know what it would be in C++ land, but in Swift I'd say we should be a computed property. Could be a function if needed.

Overall, having two state vars that need to stay in sync - with no guarantee that future maintainers will always remember that - seems like opening ourselves up to future bugs.

Reply:
I made a comment about this below, but I see this discussion is related. I'd say it's confusing to have On and Off be part of the ConnectionState because they are outside of the scope of that enum as it's described here. How about, when there are no connection tasks ongoing you either make that state a `nullptr` or introduce a `ConnectionState::Idle` option?

Reply:
I don't feel strongly on whether we include an On/Off, but I am concerned about having two states that are closely connected, yet it's technically possible to modify just one.

Reply:
> when there are no connection tasks ongoing you either make that state a nullptr or introduce a ConnectionState::Idle option?

@brizental I don't think this will work given that a VPN could be toggled on (`ConnectionStateIdle`) and be in a secondary state such as `InternetProbe` or `ServerProbe`. I'm starting to think that it would make more sense for the VPN on/off to simply be a bool value when on is true and off is false.

Reply:
Are you suggesting that the VPNClientState would be a boolean? Or is that a different variable?

Reply:
@mcleinman That would be a different variable (bool) in the controller, set to true when the VPN is toggled on, and false otherwise.

Reply:
> @brizental I don't think this will work given that a VPN could be toggled on (ConnectionStateIdle) and be in a secondary state such as InternetProbe or ServerProbe.

Hm, but in this case it would not be idle, it would be performing internet/server probe task.

Reply:
@mcleinman No, I'm proposing we'd have one state machine `ConnectionState` in Connection Manager and one bool in the Controller to keep track of whether or not the VPN is on.

@brizental I see. I think it makes sense to have an Idle state for when we aren't running any checks in Connection Manager, and a bool to keep track of VPN on/off as explained above. Does that sound good?

Reply:
I think it sounds good yes.

Reply:
Sounds good, thanks!